### PR TITLE
Implement str.isdigit

### PR DESCRIPTION
**Pull request by Rubtsowa**

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR. It's currently failing flake8 https://dev.azure.com/numba/numba/_build/results?buildId=3161&view=logs&jobId=687448c2-4012-5c9e-43b3-3520c9635a30&taskId=dba03d4a-57a7-5e59-a47a-79fe530f297a&lineStart=64&lineEnd=65&colStart=1&colEnd=1

File: numba/tests/test_unicode.py [hunk 1]
```diff
@@ -1332,6 +1332,25 @@ def test_lower(self):
...
+        cpython = ['\u2460', '\xbc', '\u0660', '\U00010401', '\U00010427',
+                   '\U00010429', '\U0001044E', '\U0001F40D', '\U0001F46F',
+                   '\U00011065', '\U0001D7F6', '\U00011066', '\U000104A0',
+                   '\U0001F107']
+        # https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Lib/test/test_unicode.py#L742-L749    # noqa: E501
+        cpython_extras = ['\uD800', '\uDFFF', '\uD800\uD800', '\uDFFF\uDFFF',
+                          'a\uD800b\uDFFF', 'a\uDFFFb\uD800',
+                          'a\uD800b\uDFFFa', 'a\uDFFFb\uD800a']
+
+        msg = 'Results of "{}".isspace() must be equal'
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]: This is "isdigit".

File: numba/unicode.py [hunk 2]
```diff
@@ -1361,6 +1361,25 @@ def impl(a):
     return impl
 
 
+# https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12056-L12085    # noqa: E501
+@overload_method(types.UnicodeType, 'isdigit')
+def unicode_isdigit(data):
+    """Implements UnicodeType.isdigit()"""
+
+    def impl(data):
+        length = len(data)
```
Comment (stuartarchibald) [inline numba/unicode.py]: Any merit in implementing this https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodeobject.c#L12070-L12074

I'm starting to think that these "single character string" branches are actually optimisations for the case that's most encountered in practice. I think the CPython impls deals with the cases in order of frequency of occurrence mixed with cost, i.e. 1 character (most likely), then empty string (cheap), then general case.

File: numba/unicode_support.py [hunk 3]
```diff
@@ -207,12 +207,10 @@ def _PyUnicode_IsDecimalDigit(ch):
 
 @register_jitable
 def _PyUnicode_ToDigit(ch):
-    raise NotImplementedError
-
-
-@register_jitable
-def _PyUnicode_IsDigit(ch):
-    raise NotImplementedError
+    ctype = _PyUnicode_gettyperecord(ch)
```
Comment (stuartarchibald) [inline numba/unicode_support.py]: This needs a github source reference with specific SHA referring to the `ToDigit` implementation.

File: numba/unicode_support.py [hunk 4]
```diff
@@ -311,6 +309,14 @@ def _PyUnicode_IsCaseIgnorable(ch):
     return ctype.flags & _PyUnicode_TyperecordMasks.CASE_IGNORABLE_MASK != 0
 
 
+# From: https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodectype.c#L123-L135    # noqa: E501
```
Comment (stuartarchibald) [inline numba/unicode_support.py]:
```suggestion
# From: https://github.com/python/cpython/blob/1d4b6ba19466aba0eb91c4ba01ba509acf18c723/Objects/unicodectype.c#L130-L135    # noqa: E501
```
the referenced line count is too large at present.

**Comment by stuartarchibald** [comment]
Please could you resolve the conflicts by merging master into this branch, thanks.

**Comment by Rubtsowa** [comment]
Conflicts resolved. Do you have anymore comment or recommendations related to this PR?

**Comment by Rubtsowa** [comment]
Сhange the label, please.

**Comment by esc** [comment]
@Rubtsowa thanks for pinging us on this PR, I have updated the label accordingly.

For future reference, it would accelerate the review process if you could write somewhat more meaningful commit messages. For example, in this pull-request, all the commits have the same message: `Implement str.isdigit` which makes it hard to get a quick overview of what has been done since the last time anyone has reviewed the pull-request, since all commits appear identical. You can find much advice on the  internet on this topic, for example the following article provides some good insights:

https://chris.beams.io/posts/git-commit/

Thanks!

**Comment by stuartarchibald** [comment]
Please could you resolve the conflicts by merging master into this branch. Thanks!

**Comment by Rubtsowa** [comment]
@stuartarchibald, I resolved conflicts.

**Comment by Rubtsowa** [comment]
Please check this PR! If everything is correct, please merge.

**Pull request closed without merging**